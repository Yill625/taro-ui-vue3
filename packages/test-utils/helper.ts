import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}