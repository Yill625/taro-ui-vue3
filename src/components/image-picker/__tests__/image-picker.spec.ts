import { mountFactory, Slots } from '@/tests/helper'
import AtImagePicker from '../index'
import Taro from '@tarojs/taro'
import { ref, unref } from 'vue'

const factory = (
  props: object = {},
  slots?: Slots
) => {
  return mountFactory(AtImagePicker, undefined, props, slots)
}

const files = ref([
  {
    url: 'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
    id: '2121',
    status: 'uploading',
  },
  {
    url: 'https://zos.alipayobjects.com/rmsportal/hqQWgTXdrlmVVYi.jpeg',
    id: '2122',
    status: 'failed',
  },
  {
    url: 'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
    id: '2121',
    status: 'uploading',
    message: '正在上传自定义消息'
  },
  {
    url: 'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
    id: '2121',
    status: 'failed',
    message: '上传失败自定义消息'
  },
  {
    url: 'https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg',
    id: '2121',
  },
])

describe('AtImagePicker', () => {
  let getEnv
  window.URL.revokeObjectURL = jest.fn()

  beforeEach(() => {
    jest.mock('@tarojs/taro')
    getEnv = jest.spyOn(Taro, 'getEnv').mockImplementation(() => {
      return Taro.ENV_TYPE.WEAPP
    })
  })

  afterEach(() => {
    getEnv.mockClear()
  })

  it('should render all nodes and match snapshot', async () => {
    const wrapper = factory({ files })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render prop class', async () => {
    const wrapper = factory({ class: 'test' })
    expect(wrapper.find('.at-image-picker').classes()).toContain('test')
  })

  it('should render prop style', async () => {
    const wrapper = factory({ style: 'color: red;' })
    expect(wrapper.find('.at-image-picker').attributes('style')).toEqual('color: red;')
  })

  it('should render prop files', async () => {
    const wrapper = factory({ files: files })
    const imageEls = wrapper.findAll('.at-image-picker__preview-img')
    const uploadingIconEls = wrapper.findAll('.at-loading')
    const failedIconEls = wrapper.findAll('.at-image-picker__status-icon--failed')
    const statusMessageEls = wrapper.findAll('.at-image-picker__status-message')
    expect(imageEls.length).toBe(unref(files).length)
    expect(uploadingIconEls.length).toBe(unref(files).filter(({ status }) => status === 'uploading').length)
    expect(failedIconEls.length).toBe(unref(files).filter(({ status }) => status === 'failed').length)
    expect(statusMessageEls.length).toBe(unref(files).filter(({ message }) => !!message).length)
    imageEls.forEach((el, index) => {
      expect(el.attributes('src')).toEqual(unref(files)[index].url)
      const statusMessageEl = el.find('.at-image-picker__status-message')
      statusMessageEl.exists() && expect(statusMessageEl.text()).toBe(unref(files)[index].message)
    })
  })

  it('should render prop length', async () => {
    let length = 5
    let row = Math.ceil((unref(files).length + 1) / length)
    const wrapper = factory({ files: files, length })
    expect(wrapper.findAll('.at-image-picker__flex-item').length).toBe(row * length)
    expect(wrapper.findAll('.at-image-picker__flex-box').length).toBe(row)

    length = 0
    row = Math.ceil((unref(files).length + 1) / 1)
    await wrapper.setProps({ length })
    await wrapper.vm.$nextTick()
    expect(wrapper.findAll('.at-image-picker__flex-item').length).toBe(row * 1)
    expect(wrapper.findAll('.at-image-picker__flex-box').length).toBe(row)
  })

  it.each([
    "scaleToFill",
    "aspectFit",
    "aspectFill",
    "widthFix",
    "top",
    "bottom",
    "center",
    "left",
    "right",
    "top left",
    "top right",
    "bottom left",
    "bottom right"
  ])('should render prop mode -- %s', async (modeOption) => {
    const wrapper = factory({ files: files, mode: modeOption })
    wrapper.findAll('.at-image-picker__preview-img').forEach(imageEl => {
      expect(imageEl.attributes('mode')).toBe(modeOption)
    })
  })

  it('should render prop showAddBtn', async () => {
    const wrapper = factory()
    expect(wrapper.find('.at-image-picker__choose-btn').exists()).toBe(true)

    await wrapper.setProps({ showAddBtn: false })
    expect(wrapper.find('.at-image-picker__choose-btn').exists()).toBe(false)
  })

})

describe('AtImagePicker Behaviours', () => {
  let getEnv
  beforeEach(() => {
    jest.mock('@tarojs/taro')
    getEnv = jest.spyOn(Taro, 'getEnv').mockImplementation(() => {
      return Taro.ENV_TYPE.WEAPP
    })
  })

  afterEach(() => {
    getEnv.mockClear()
  })

  it('should trigger onChange by clicking remove button', async () => {
    const onChange = jest.fn()
    const wrapper = factory({ files: files, onChange: onChange })
    expect(wrapper.findAll('.at-image-picker__preview-img').length).toBe(5)

    await wrapper.find('.at-image-picker__remove-btn').trigger('tap')
    expect(onChange).toBeCalled()
    expect(wrapper.findAll('.at-image-picker__preview-img').length).toBe(5)
  })

  it('should trigger onChange by clicking add button', async () => {
    const spy = jest.spyOn(Taro, 'chooseImage').mockImplementation(() => {
      return new Promise((resolve) => {
        resolve({
          tempFilePaths: [''],
          tempFiles: [{ path: '', size: 0 }],
          errMsg: ''
        })
      })
    })

    const onChange = jest.fn()
    const wrapper = factory({ files: files, onChange: onChange, count: 9, multiple: true })
    expect(wrapper.findAll('.at-image-picker__preview-img').length).toBe(5)

    await wrapper.find('.at-image-picker__choose-btn').trigger('tap')
    expect(onChange).toBeCalled()

    spy.mockClear()
  })

  it('should trigger onImageClick', async () => {
    const onImageClick = jest.fn()
    const wrapper = factory({ files: files, onImageClick: onImageClick })

    await wrapper.find('.at-image-picker__preview-img').trigger('tap')
    expect(onImageClick).toBeCalled()
  })

  it('should trigger onFail if chooseImage fails', async () => {
    const spy = jest.spyOn(Taro, 'chooseImage').mockImplementation(() => {
      return new Promise((resolve, reject) => {
        reject({ errMsg: 'failed' })
      })
    })

    const onFail = jest.fn()
    const wrapper = factory({ files: files, onFail: onFail })

    await wrapper.find('.at-image-picker__choose-btn').trigger('tap')
    await wrapper.vm.$nextTick()
    expect(onFail).toBeCalled()
    spy.mockClear()
  })

  test('modify files to show uploading status', async () => {
    const index = 4
    const status = 'uploading'
    const message = '正在上传'
    const wrapper = factory({ files: files })
    unref(files)[index].status = status
    unref(files)[index].message = message
    await wrapper.vm.$nextTick()
    const el = wrapper.findAll('.at-image-picker__item')[index]
    expect(el.find('.at-loading').exists()).toBeTruthy()
    expect(el.find('.at-image-picker__status-message').text()).toBe(message)
  })

  test('modify files to show failed status', async () => {
    const index = 4
    const status = 'failed'
    const message = '上传失败'
    const wrapper = factory({ files: files })
    unref(files)[index].status = status
    unref(files)[index].message = message
    await wrapper.vm.$nextTick()
    const el = wrapper.findAll('.at-image-picker__item')[index]
    el.get('.at-image-picker__status-icon--failed')
    expect(el.find('.at-image-picker__status-message').text()).toBe(message)
  })

  it.each([
    {
      index: 4,
      status: 'done',
      message: '不会显示的message'
    },
    {
      index: 4,
      status: '',
      message: '不会显示的message'
    },
    {
      index: 4,
      message: '不会显示的message'
    },
    {
      index: 4,
    }
  ])('modify files to show done status ', async ({ index, status, message }) => {
    const wrapper = factory({ files: files })
    unref(files)[index].status = status
    unref(files)[index].message = message
    await wrapper.vm.$nextTick()
    const el = wrapper.findAll('.at-image-picker__item')[index]
    expect(() => el.get('.at-image-picker__upload-status')).toThrowError()
  })

})
