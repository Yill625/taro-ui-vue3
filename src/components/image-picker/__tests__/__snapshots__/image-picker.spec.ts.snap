// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtImagePicker should render all nodes and match snapshot 1`] = `
<view class="at-image-picker">
  <view class="at-image-picker__flex-box">
    <view class="at-image-picker__flex-item">
      <view class="at-image-picker__item">
        <view class="at-image-picker__remove-btn"></view>
        <image class="at-image-picker__preview-img" mode="aspectFill" src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"></image>
        <view class="at-image-picker__upload-status">
          <view class="at-loading">
            <view class="at-loading__ring" style="border: 1px solid #fff; border-color: #fff transparent transparent transparent;"></view>
            <view class="at-loading__ring" style="border: 1px solid #fff; border-color: #fff transparent transparent transparent;"></view>
            <view class="at-loading__ring" style="border: 1px solid #fff; border-color: #fff transparent transparent transparent;"></view>
          </view>
        </view>
      </view>
    </view>
    <view class="at-image-picker__flex-item">
      <view class="at-image-picker__item">
        <view class="at-image-picker__remove-btn"></view>
        <image class="at-image-picker__preview-img" mode="aspectFill" src="https://zos.alipayobjects.com/rmsportal/hqQWgTXdrlmVVYi.jpeg"></image>
        <view class="at-image-picker__upload-status">
          <view class="at-image-picker__status-icon at-image-picker__status-icon--failed"></view>
        </view>
      </view>
    </view>
    <view class="at-image-picker__flex-item">
      <view class="at-image-picker__item">
        <view class="at-image-picker__remove-btn"></view>
        <image class="at-image-picker__preview-img" mode="aspectFill" src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"></image>
        <view class="at-image-picker__upload-status">
          <view class="at-loading">
            <view class="at-loading__ring" style="border: 1px solid #fff; border-color: #fff transparent transparent transparent;"></view>
            <view class="at-loading__ring" style="border: 1px solid #fff; border-color: #fff transparent transparent transparent;"></view>
            <view class="at-loading__ring" style="border: 1px solid #fff; border-color: #fff transparent transparent transparent;"></view>
          </view>
          <view class="at-image-picker__status-message at-image-picker__status-message--uploading">正在上传自定义消息</view>
        </view>
      </view>
    </view>
    <view class="at-image-picker__flex-item">
      <view class="at-image-picker__item">
        <view class="at-image-picker__remove-btn"></view>
        <image class="at-image-picker__preview-img" mode="aspectFill" src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"></image>
        <view class="at-image-picker__upload-status">
          <view class="at-image-picker__status-icon at-image-picker__status-icon--failed"></view>
          <view class="at-image-picker__status-message at-image-picker__status-message--failed">上传失败自定义消息</view>
        </view>
      </view>
    </view>
  </view>
  <view class="at-image-picker__flex-box">
    <view class="at-image-picker__flex-item">
      <view class="at-image-picker__item">
        <view class="at-image-picker__remove-btn"></view>
        <image class="at-image-picker__preview-img" mode="aspectFill" src="https://zos.alipayobjects.com/rmsportal/PZUUCKTRIHWiZSY.jpeg"></image>
      </view>
    </view>
    <view class="at-image-picker__flex-item">
      <view class="at-image-picker__item at-image-picker__choose-btn">
        <view class="add-bar"></view>
        <view class="add-bar"></view>
      </view>
    </view>
    <view class="at-image-picker__flex-item">
      <!---->
    </view>
    <view class="at-image-picker__flex-item">
      <!---->
    </view>
  </view>
</view>
`;
